include/mata/nfa/delta.hh

```diff
@@ -121,64 +121,105 @@ public:
-     *  we iterate.
+     * @brief Iterator over moves represented as @c Move instances.
+     *
+     * It iterates over pairs (symbol, target) for the given @c StatePost.
      */
-    struct moves_const_iterator {
-    private:
-        const std::vector<SymbolPost>& symbol_posts_{};
-        std::vector<SymbolPost>::const_iterator symbol_post_it_{};
-        StateSet::const_iterator target_states_it_{};
-        bool is_end_{ false };
-        Move move_{};
-
-    public:
-        using iterator_category = std::forward_iterator_tag;
-        using value_type = Move;
-        using difference_type = unsigned;
-        using pointer = Move*;
-        using reference = Move&;
-
-        explicit moves_const_iterator(const std::vector<SymbolPost>& symbol_posts, bool is_end = false);
-
-        moves_const_iterator(const std::vector<SymbolPost>& symbol_posts,
-                             std::vector<SymbolPost>::const_iterator symbol_posts_it,
-                             StateSet::const_iterator target_states_it, bool is_end = false);
-
-        moves_const_iterator(const moves_const_iterator& other) = default;
-
-        const Move& operator*() const { return move_; }
-
-        // Prefix increment
-        moves_const_iterator& operator++();
-        // Postfix increment
-        const moves_const_iterator operator++(int);
-
-        moves_const_iterator& operator=(const moves_const_iterator& x);
-
-        bool operator==(const moves_const_iterator& other) const;
-        bool operator!=(const moves_const_iterator& other) const { return !(*this == other); };
-    };
-
-    moves_const_iterator moves_cbegin() const { return moves_const_iterator(ToVector()); }
-    moves_const_iterator moves_cend() const { return moves_const_iterator(ToVector(), true); }
-    moves_const_iterator moves_begin() const { return moves_cbegin(); }
-    moves_const_iterator moves_end() const { return moves_cend(); }
-
     class Moves {
     public:
-        moves_const_iterator begin_;
-        moves_const_iterator end_;
-        moves_const_iterator begin() const { return begin_; }
-        moves_const_iterator end() const { return end_; }
-    };
+        /**
+         * @brief Whether to look up first symbol to iterate over from the beginning or from the end of @c StatePost
+         *  iterator.
+
+         * Iterates over symbol posts with symbols in interval [first_symbol_, last_symbol_].
+         */
+        enum class FirstSymbolLookupDiretion {
+            Forward, ///< Lookup from the beginning to the end.
+            Backward ///< Lookup from the end to the beginning.
+        };
+
+        /**
+         * Iterator over moves.
+         */
+        class const_iterator {
+        private:
+            const StatePost* state_post_{ nullptr };
+            StatePost::const_iterator state_post_it_{};
+            StateSet::const_iterator symbol_post_it_{};
+            Symbol last_symbol_{ Limits::max_symbol };
```

Comment:
I wonder if `sentinel_` would be appropriate name? It's short, elegant, well-established.

Reply:
In act I would not know what sentinel means.

Reply:
I believe, it is quite well-established naming for our `zarážka` in programming (e.g. in some lookup algorithm, I am fairly sure it is used). `last_symbol` sounds like something at `.back()`, which I'm not sure if applies for this particular case.

Reply:
To my best knowledge, it is as @tfiedor says. However, in this case I meant to truly use the notion of first symbol and last symbol to explicitly state that we iterate over a closed interval `[fist_symbol, last_symbol]`, that is, both first symbol and last symbol are iterated over. Sentinel is what we usually use for one after the last element, it is basically what we call the end iterator in C++ iterator semantics.

Reply:
How about just `max_symbol` then?

Reply:
My issue is, that it is little ambiguous (could be last as in position, or last as last seen, or last as upper limit, etc.)

Reply:
`max_symbol` is fine with me. When talking about intervals, you use `[first, last]`, however, instead of `[first, max]`, no?

But I agree that in order to avoid ambiguous code, `max` might be easier to work with.

Reply:
For me `max` is more ambigious than `last`, especially when it is initialized to `Limits::max_symbol`. I think a comment here would be good.

Reply:
So what about keeping `last_symbol`, but adding an explanation, that it is by default the maximal symbol and that it means the last symbol in the vector, not the last + 1. Would that be sufficient?

Reply:
I added more explanations. Hopefully, the distinction will now be clear.